## Objective
Discover factors on hotel booking cancellation and provide recommendations to increase revenue. 

## The Data

They have provided you with their bookings data in a file called `hotel_bookings.csv`, which contains the following:

| Column     | Description              |
|------------|--------------------------|
| `Booking_ID` | Unique identifier of the booking. |
| `no_of_adults` | The number of adults. |
| `no_of_children` | The number of children. |
| `no_of_weekend_nights` | Number of weekend nights (Saturday or Sunday). |
| `no_of_week_nights` | Number of week nights (Monday to Friday). |
| `type_of_meal_plan` | Type of meal plan included in the booking. |
| `required_car_parking_space` | Whether a car parking space is required. |
| `room_type_reserved` | The type of room reserved. |
| `lead_time` | Number of days before the arrival date the booking was made. |
| `arrival_year` | Year of arrival. |
| `arrival_month` | Month of arrival. |
| `arrival_date` | Date of the month for arrival. |
| `market_segment_type` | How the booking was made. |
| `repeated_guest` | Whether the guest has previously stayed at the hotel. |
| `no_of_previous_cancellations` | Number of previous cancellations. |
| `no_of_previous_bookings_not_canceled` | Number of previous bookings that were canceled. |
| `avg_price_per_room` | Average price per day of the booking. |
| `no_of_special_requests` | Count of special requests made as part of the booking. |
| `booking_status` | Whether the booking was cancelled or not. |

Source (data has been modified): https://www.kaggle.com/datasets/ahsan81/hotel-reservations-classification-dataset

## Visualization
Visualization of data is done by tableau after cleaning data with python. Click [here](https://public.tableau.com/app/profile/max8645/viz/Hotelanalysis_16824894033860/Story1#1) to view interactive dashboards. Start from identy target to cancel weekday.

## Findings
The hotel has a 32.74% cancellation rate on its entire customer population (32780), which causes 1.18M financial loss.

For customers who canceled hotel bookings, most of them are non-repeated customers. Customers with children have a minor impact since it only dominates 7.53% in this group disregards the number of children. Customer with size 2 without children has a significant impact on hotel booking cancellation, which rules 66.88% and makes 63.87% revenue loss.

As the number of children and adults increases, expecting the average room price to be more expensive. However, there is an outlier in the cancel booking group. The customer with size 2 has 9 children, the average room price is only $76.5.

The given data is from 2017 and 2018. For data in 2017, it is only accessible begins from September up to December. Evaluating average room price impacts on customer size in cancellation on each month from September to December in both years, the average room price reaches its peak in September. The average room price starts to decline in the following months. However, when the average room price is at its maximum point, the customer size reaches its peak in October. This is an interesting fact that price drops and demand decreases. This is likely due to the coming holiday season, as customers wait for better prices.

For customers with two people without children, they prefer room type 1 the most, followed by room type 4 and room type 2 respectively. The most favorite meal plans are meal plan 1, then meal plan 2. Interestingly, customers in this group cancel their bookings most likely happens on Thursday and Wednesday based on the data in 2017 and 2018 respectively.

## Conclusion and Reconmendation
To conclude, the hotel should focus on the customer with size two without kids since it is the main portion in canceling hotel bookings. Customers are sensitive to the average room price due to the coming holiday season. Therefore, the adjusted monthly average room price should consider the impact of the holiday season. Changing average room price with a small margin monthly such that the change of price margin matches the discount rate during the holiday season.  Additionally, based on the pattern of canceling hotel bookings, it is better to provide promotions on Thursdays and Wednesdays to encourage customers to stay with the hotel.  For example, special days in a week with a bundle(room type 4 and meal plan 1) to attract more customers or low down customer churn rate.



In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pylab as plt

In [2]:
hotels = pd.read_csv("data/hotel_bookings.csv")
hotels.head(10)

Booking_ID  no_of_adults  ...  no_of_special_requests  booking_status
0   INN00001           NaN  ...                     NaN    Not_Canceled
1   INN00002           2.0  ...                     1.0    Not_Canceled
2   INN00003           1.0  ...                     0.0        Canceled
3   INN00004           2.0  ...                     0.0        Canceled
4   INN00005           2.0  ...                     0.0        Canceled
5   INN00006           2.0  ...                     1.0        Canceled
6   INN00007           2.0  ...                     1.0    Not_Canceled
7   INN00008           2.0  ...                     1.0    Not_Canceled
8   INN00009           3.0  ...                     1.0    Not_Canceled
9   INN00010           2.0  ...                     3.0    Not_Canceled

[10 rows x 19 columns]

In [3]:
hotels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 19 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Booking_ID                            36275 non-null  object 
 1   no_of_adults                          35862 non-null  float64
 2   no_of_children                        35951 non-null  float64
 3   no_of_weekend_nights                  35908 non-null  float64
 4   no_of_week_nights                     35468 non-null  float64
 5   type_of_meal_plan                     35749 non-null  object 
 6   required_car_parking_space            33683 non-null  float64
 7   room_type_reserved                    35104 non-null  object 
 8   lead_time                             35803 non-null  float64
 9   arrival_year                          35897 non-null  float64
 10  arrival_month                         35771 non-null  float64
 11  arrival_date   

In [4]:
# get the portion of missing values in each columns
missing_por = np.round(hotels.isnull().mean()*100,2)
missing_por

Booking_ID                              0.00
no_of_adults                            1.14
no_of_children                          0.89
no_of_weekend_nights                    1.01
no_of_week_nights                       2.22
type_of_meal_plan                       1.45
required_car_parking_space              7.15
room_type_reserved                      3.23
lead_time                               1.30
arrival_year                            1.04
arrival_month                           1.39
arrival_date                            2.70
market_segment_type                     4.17
repeated_guest                          1.62
no_of_previous_cancellations            1.37
no_of_previous_bookings_not_canceled    1.52
avg_price_per_room                      1.27
no_of_special_requests                  2.18
booking_status                          0.00
dtype: float64

In [5]:
hotels.describe()

no_of_adults  no_of_children  ...  avg_price_per_room  no_of_special_requests
count  35862.000000    35951.000000  ...        35815.000000            35486.000000
mean       1.845017        0.105366  ...          103.418207                0.619343
std        0.518652        0.402871  ...           35.057342                0.785849
min        0.000000        0.000000  ...            0.000000                0.000000
25%        2.000000        0.000000  ...           80.300000                0.000000
50%        2.000000        0.000000  ...           99.450000                0.000000
75%        2.000000        0.000000  ...          120.000000                1.000000
max        4.000000       10.000000  ...          540.000000                5.000000

[8 rows x 14 columns]

The main challenging for missing value in this dataset is to fill missing values in the timeframe(no_of_weekend_nights, no_of_week_nights ,lead_time, arrival_year, arrival_month and arrival_date) and price(avg_price_per_room). These factors on the timeframe has significantly impact on analysis to understand the behavior of consumers booking hotels.

Since the portion of missing values in no_of_weekend_nights(1.01%) ,no_of_week_nights(2.22%) ,lead_time(1.30%), arrival_year(1.04%), arrival_month(1.394%),arrival_date(2.70%) and avg_price_per_room(1.27%) are less than 3% percent on the entire given dataset, we will drop the missing values in these columns

In [6]:
# dropping 'lead_time','arrival_year','arrival_month', 'arrival_date', 'avg_price_per_room'
hotel_update = hotels.dropna(subset=['no_of_weekend_nights','no_of_week_nights','lead_time','arrival_year','arrival_month', 'arrival_date', 'avg_price_per_room'])

hotel_update.shape[0]/hotels.shape[0]

0.9041212956581668

In [7]:
np.round(hotel_update.isnull().mean()*100,2)

Booking_ID                              0.00
no_of_adults                            0.86
no_of_children                          0.66
no_of_weekend_nights                    0.00
no_of_week_nights                       0.00
type_of_meal_plan                       1.30
required_car_parking_space              6.49
room_type_reserved                      3.22
lead_time                               0.00
arrival_year                            0.00
arrival_month                           0.00
arrival_date                            0.00
market_segment_type                     2.18
repeated_guest                          1.58
no_of_previous_cancellations            1.36
no_of_previous_bookings_not_canceled    0.86
avg_price_per_room                      0.00
no_of_special_requests                  2.13
booking_status                          0.00
dtype: float64

In [8]:
hotel_update.describe()

no_of_adults  no_of_children  ...  avg_price_per_room  no_of_special_requests
count  32516.000000    32579.000000  ...        32797.000000            32098.000000
mean       1.845215        0.105190  ...          103.453105                0.620942
std        0.518457        0.403592  ...           35.041171                0.786705
min        0.000000        0.000000  ...            0.000000                0.000000
25%        2.000000        0.000000  ...           80.390000                0.000000
50%        2.000000        0.000000  ...           99.450000                0.000000
75%        2.000000        0.000000  ...          120.000000                1.000000
max        4.000000       10.000000  ...          540.000000                5.000000

[8 rows x 14 columns]

In [9]:
#no_of_adults fill no_of_adults with most frequent number
hotel_update['no_of_adults'] =  hotel_update['no_of_adults'].fillna(hotel_update['no_of_adults'].mode().loc[0])

# check if there is invalide value in no_of_adults which is less than 1
print("Number of adult less than 1?",(hotel_update['no_of_adults'] < 1).any())

Number of adult less than 1? True


In [10]:
# update the value in most frequent number
mask_adult = hotel_update['no_of_adults'] < 1
hotel_update.loc[mask_adult,'no_of_adults'] = hotel_update['no_of_adults'].mode().loc[0]

In [11]:
# fill missing value with the most frequent number
hotel_update['no_of_children'] =  hotel_update['no_of_children'].fillna(hotel_update['no_of_children'].mode().loc[0])

In [12]:
# fill required_car_parking_space with most frequent number
hotel_update['required_car_parking_space'] = hotel_update['required_car_parking_space'].fillna(hotel_update['required_car_parking_space'].mode().loc[0])


In [13]:
# repeated_guest depends has two types for 0, it does not have no_of_previous_cancellations neither no_of_previous_bookings_not_canceled 
cancel_vars = ['no_of_previous_cancellations','no_of_previous_bookings_not_canceled']
mask_rg = hotel_update['repeated_guest'] == 0 

#has missing values?
print("Missing values: ", hotel_update[mask_rg][cancel_vars].isnull().any(), sep='\n')

#set 0 on no_of_previous_cancellations & no_of_previous_cancellations when repeated_guest is 0
hotel_update.loc[mask_rg,cancel_vars] = 0

# fill repeated_guest is null as 0 when no_of_previous_cancellations & no_of_previous_cancellations are 0
mask_0 = (hotel_update['no_of_previous_cancellations'] == 0) & (hotel_update['no_of_previous_bookings_not_canceled'] ==0) &(hotel_update['repeated_guest'].isnull())
hotel_update.loc[mask_0,'repeated_guest'] = 0

print("Missing values: ", hotel_update[mask_rg][cancel_vars].isnull().any(), sep='\n')

Missing values: 
no_of_previous_cancellations            True
no_of_previous_bookings_not_canceled    True
dtype: bool
Missing values: 
no_of_previous_cancellations            False
no_of_previous_bookings_not_canceled    False
dtype: bool


In [14]:
#following the same steps for repeat_guest = 1
msk_rg = hotel_update['repeated_guest'] == 1
msk_1 = (hotel_update['no_of_previous_cancellations'].isnull()) | (hotel_update['no_of_previous_bookings_not_canceled'].isnull())
#has missing values?
print("Missing values: ", hotel_update[msk_rg][cancel_vars].isnull().any(), sep='\n')

Missing values: 
no_of_previous_cancellations            True
no_of_previous_bookings_not_canceled    True
dtype: bool


In [15]:
hotel_update.loc[(msk_rg&msk_1)]

Booking_ID  no_of_adults  ...  no_of_special_requests  booking_status
4015    INN04016           1.0  ...                     0.0    Not_Canceled
6278    INN06279           1.0  ...                     1.0    Not_Canceled
8176    INN08177           1.0  ...                     0.0    Not_Canceled
8250    INN08251           1.0  ...                     0.0    Not_Canceled
19734   INN19735           1.0  ...                     NaN    Not_Canceled
33396   INN33397           1.0  ...                     NaN    Not_Canceled

[6 rows x 19 columns]

In [16]:
# update missing values
hotel_update.loc[(msk_rg&msk_1), 'no_of_previous_cancellations'] = 0
hotel_update.loc[(msk_rg&msk_1), 'no_of_previous_bookings_not_canceled'] = 1

print("Missing values: ", hotel_update[msk_rg][cancel_vars].isnull().any(), sep='\n')

Missing values: 
no_of_previous_cancellations            False
no_of_previous_bookings_not_canceled    False
dtype: bool


In [17]:
# dropping repeat_guest OR no_of_previous_cancellations is null WHEN repeat_guest is null
mask_null = hotel_update['repeated_guest'].isnull()
mask_nulls = (mask_null & hotel_update['no_of_previous_cancellations'].isnull()) | (mask_null & hotel_update['no_of_previous_bookings_not_canceled'].isnull())
hotel_update.drop(hotel_update[mask_nulls].index, inplace=True)
#fill the missing value in repeat_guess when repeat_guest OR no_of_previous_cancellations is not zero.
hotel_update.loc[mask_null,'repeated_guest'] = 1

In [18]:
# set missing values to be zero
mask_rqst = hotel_update['no_of_special_requests'].isnull()
hotel_update.loc[mask_rqst, 'no_of_special_requests']= 0

In [19]:
# fill the missing value of type_of_meal_plan, room_type_reserved, market_segment_type with most frequent value
hotel_update['type_of_meal_plan'] = hotel_update['type_of_meal_plan'].fillna(hotel_update['type_of_meal_plan'].mode()[0])
hotel_update['room_type_reserved'] = hotel_update['room_type_reserved'].fillna(hotel_update['room_type_reserved'].mode()[0])
hotel_update['market_segment_type'] = hotel_update['market_segment_type'].fillna(hotel_update['market_segment_type'].mode()[0])

In [20]:
hotel_update.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32780 entries, 1 to 36274
Data columns (total 19 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Booking_ID                            32780 non-null  object 
 1   no_of_adults                          32780 non-null  float64
 2   no_of_children                        32780 non-null  float64
 3   no_of_weekend_nights                  32780 non-null  float64
 4   no_of_week_nights                     32780 non-null  float64
 5   type_of_meal_plan                     32780 non-null  object 
 6   required_car_parking_space            32780 non-null  float64
 7   room_type_reserved                    32780 non-null  object 
 8   lead_time                             32780 non-null  float64
 9   arrival_year                          32780 non-null  float64
 10  arrival_month                         32780 non-null  float64
 11  arrival_date   

In [21]:
# used portion of data after cleaning
print('data size after cleaning in the old data: ', np.round(100* hotel_update.shape[0]/hotels.shape[0], 2))

# update float64 into int and keep avg_price_per_room as float
cols = list(hotel_update.select_dtypes(include='float64').columns)
cols.remove('avg_price_per_room')
hotel_update[cols] = hotel_update[cols].astype('int')

data size after cleaning in the old data:  90.37


In [22]:
hotel_update.describe()

no_of_adults  no_of_children  ...  avg_price_per_room  no_of_special_requests
count  32780.000000    32780.000000  ...        32780.000000            32780.000000
mean       1.854240        0.104545  ...          103.454150                0.607657
std        0.503793        0.402436  ...           35.038711                0.783363
min        1.000000        0.000000  ...            0.000000                0.000000
25%        2.000000        0.000000  ...           80.390000                0.000000
50%        2.000000        0.000000  ...           99.450000                0.000000
75%        2.000000        0.000000  ...          120.000000                1.000000
max        4.000000       10.000000  ...          540.000000                5.000000

[8 rows x 14 columns]

In [49]:
hotel_update.to_csv('hotel_data.csv', index='False')